### PR TITLE
Display snapshot timer

### PR DESCRIPTION
**Pull request by GuillaumeGomez**

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -407,6 +416,7 @@ fn build_ui(app: &App, application: &gtk::Application) {
     window.set_position(gtk::WindowPosition::Center);
     window.set_default_size(350, 300);
 
+    let timeout: Rc<RefCell<Option<glib::source::SourceId>>> = Rc::new(RefCell::new(None));
```
Comment (sdroege) [inline src/main.rs]: Better store this in the app struct

Comment (sdroege) [reply src/main.rs]: Then we can also disable the timer eg when going to the settings

File: src/main.rs [hunk 2]
```diff
@@ -431,12 +441,63 @@ fn build_ui(app: &App, application: &gtk::Application) {
...
+            overlay_text.set_visible(true);
+            overlay_text.set_text(&settings.timer_length.to_string());
+
+            let overlay_text_weak = overlay_text.downgrade();
+            let timeout_weak = Rc::downgrade(&timeout);
+            let source = gtk::timeout_add_seconds(1, move || {
+                let timeout = upgrade_weak!(timeout_weak, glib::Continue(false));
+                let overlay_text = upgrade_weak!(overlay_text_weak, glib::Continue(false));
+                let remaining = u32::from_str_radix(&overlay_text.get_text()
+                                                                 .unwrap_or_else(|| "1".to_owned()),
```
Comment (sdroege) [inline src/main.rs]: When would the parsing ever fail? Also seems nicer to store an u32 somewhere for this. We have an app struct :)

File: src/main.rs [hunk 3]
```diff
@@ -431,12 +441,63 @@ fn build_ui(app: &App, application: &gtk::Application) {
...
+                                                    10).unwrap_or_else(|_| 1);
+                if remaining < 2 {
```
Comment (sdroege) [inline src/main.rs]: I.e. == 1? :)

File: src/main.rs [hunk 4]
```diff
@@ -462,7 +523,9 @@ fn build_ui(app: &App, application: &gtk::Application) {
     // vertically or horizontally
     let vbox = gtk::Box::new(gtk::Orientation::Vertical, 0);
     vbox.pack_start(&view, true, true, 0);
-    window.add(&vbox);
+
+    overlay.add(&vbox);
+    window.add(&overlay);
```
Comment (sdroege) [inline src/main.rs:592]: Maybe we could only create the overlay on demand when needed?

Comment (GuillaumeGomez) [reply src/main.rs:592]: Seems simpler this way. Like that it's initiated at the beginning once and for all, and we just change its visibility.

Comment (sdroege) [reply src/main.rs:592]: OK :)

File: src/main.rs [hunk 5]
```diff
@@ -431,12 +441,63 @@ fn build_ui(app: &App, application: &gtk::Application) {
...
+
+        if let Some(t) = timeout.borrow_mut().take() {
+            glib::source::source_remove(t);
+            overlay_text.set_visible(false);
+            return
+        }
+        if settings.timer_length == 0 {
+            take_snapshot();
+        } else {
+            overlay_text.set_visible(true);
```
Comment (sdroege) [inline src/main.rs]: It goes into this branch for me but I never actually see the overlay

Comment (GuillaumeGomez) [reply src/main.rs]: Ah? That is surprising... It works well for me (on linux mint).

Comment (GuillaumeGomez) [reply src/main.rs]: Just set it visible all the time. Like that you'll know for sure it's not something that isn't working.

**Comment by GuillaumeGomez** [comment]
Moved things into `AppInner`.

File: src/main.rs (hunk 5 above)
Comment (sdroege) [reply src/main.rs]: Doesn't appear then either :)

Comment (GuillaumeGomez) [reply src/main.rs]: ... Do the overlay example works for you? (the C one in GNOME example repository)

Comment (sdroege) [reply src/main.rs]: The ones from `gtk3-demo`? Yes those work for me. I'll debug this later :)

Linux Mint means no gnome-shell and no Wayland I assume?

Comment (GuillaumeGomez) [reply src/main.rs]: There is gnome-shell apparently but not Wayland (that might it?).

**Comment by sdroege** [comment]
@GuillaumeGomez Did some changes, please check if this also works for you.

The CSS usage you did there was not really supported and only worked by accident in older GTK versions. That took a while to figure out :)

**Comment by sdroege** [comment]
I'd also like to refactor things into more functions tomorrow

File: src/main.rs [hunk 6]
```diff
@@ -307,9 +306,11 @@ fn build_snapshot_settings_window(parent: &gtk::Window) {
         if let Some(parent) = s.parent() {
             if !parent.exists() {
                 if let Err(e) = create_dir_all(parent) {
-                    eprintln!("Error when trying to build settings folder '{}': {:?}",
```
Comment (GuillaumeGomez) [inline src/main.rs:291]: You're part of those kind of indenters. I see. :p

**Comment by GuillaumeGomez** [comment]
Damn, wouldn't have thought to that... I'll check tomorrow.

File: src/main.rs (hunk 6 above)
Comment (sdroege) [reply src/main.rs:291]: rustfmt :) I don't think about these things

**Pull request merged**